### DISCUSSION TITLE
medium post feedback

### DISCUSSION THREAD

**Discussion by chunhtai** [aliyazdi75/gallery#5] [General]
I took a brief look at the https://medium.com/@aliyazdi75/bc5953251c3e I think This document is well written and touches almost all the functionalities of router API. It also reminds me of the nested routing scenario we are trying to clarify https://github.com/flutter/uxr/issues/35, because it uses nested routers. Although I feel the use case of nested routers here seems to be a bit redundant and can probably be solved by using just one router, can you talk more about why you choose to use two routers instead of one in your example?

Also feel free to respond to https://github.com/flutter/uxr/issues/35 if you have any feedback on the subject.

Thanks for writing this amazing medium post!!

**Comment by aliyazdi75** [comment]
Thank you so much for your time to read my article.

Yes, I think the reasons to separate routers are ambiguous and I will edit that section.

The Gallery aims to represent good architecture and reduce complexity. Nested routers are good to make our goals. This project is not so big and using the nested router seems to be redundant, but I used nested to explain to the readers, how to make the nested router, thus they can understand and make them by themselves for their big project. But if I was going to explain the reason, I would say:

1. As mentioned in the flutter/uxr#35, this is useful for multiple teams working on the same project, and by structuring in a hierarchical structure way, this makes the project less complex:
For example, in the Gallery, we have two main simple screens: [`LoginPage`](https://github.com/aliyazdi75/gallery/blob/main/lib/presentation/screens/login/login.dart) and [`GalleryShell`](https://github.com/aliyazdi75/gallery/blob/main/lib/presentation/screens/gallery/gallery.dart), as they aren't related to each other, they are separated in their directory presentation layer.
The `GalleryShell` has some subpages:
   - [`RootPage`](https://github.com/aliyazdi75/gallery/blob/da68d1741f377b4820e55a1b1834a54f5952111f/lib/main.dart#L158) (the first album page),
   - [`AlbumPage`](https://github.com/aliyazdi75/gallery/blob/main/lib/presentation/screens/album/view/album.dart)
   - [`FullscreePage`](https://github.com/aliyazdi75/gallery/blob/main/lib/presentation/screens/fullscreen/fullscreen.dart)
2. We should consider changing our states as the related changes in the `onPopPage`, so bringing all changes in one router which they aren't related to each other is not a good idea. For example, this is the [`onPopPage`](https://github.com/aliyazdi75/gallery/blob/da68d1741f377b4820e55a1b1834a54f5952111f/lib/presentation/screens/gallery/router.dart#L81) for the `GalleryShell`.
3. Like the second reason, we sometimes need to define some observers like the `HeroController`, and this observer relates to the `GalleryShell` subpages and we have to define it just for that router, for example, this is the [`observers`](https://github.com/aliyazdi75/gallery/blob/da68d1741f377b4820e55a1b1834a54f5952111f/lib/presentation/screens/gallery/router.dart#L37) for the `GalleryShell`.
4. We can simply add the specific widgets for each router that we need. For example, if we want to change the text scale of the `GalleryShell` subpages, we can simply wrap `GalleryShell` by this widget:

    ```dart
    MediaQuery(
        data: MediaQuery.of(context).copyWith(
          textScaleFactor: 4.0,
        ),
        child: GalleryShell(),
    );
    ```
    instead of wrapping every subpage with this widget.

I think these are good reasons to make the nested routers at the beginning for a big project.
